File: LeetCode/1326. Minimum Number of Taps to Open to Water a Garden/solution.py
# class Solution:
#     def minTaps(self, n: int, ranges: List[int]) -> int:
#         intervals = [[i-radius, i+radius] for i, radius in enumerate(ranges)]
#         intervals.sort()
#         dp = [float('inf')]*(n+1)
#         dp[0] = 0
#         for a, b in intervals:
#             a = max(a,0)
#             b = min(b,n)
#             for i in range(a,b+1):
#                 dp[i] = min(dp[i],dp[a]+1)
#         if dp[n] == float('inf'):
#             return -1
#         return dp[n]


class Solution:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        intervals = [[i-radius, i+radius] for i, radius in enumerate(ranges)]
        intervals.sort()
        m = len(intervals)
        cur, reach, cnt = 0, 0, 0
        while reach < n:
            tmp = reach
            if cur >= m or intervals[cur][0] > tmp:
                return -1
            while cur < m and intervals[cur][0] <= tmp:
                reach = max(reach, intervals[cur][1])
                cur += 1
            cnt += 1
        return cnt

File: LeetCode/1326. Minimum Number of Taps to Open to Water a Garden/solution.cpp
using pii = pair<int,int>;
const int inf = 0x3f3f3f3f;

class Solution {
public:
    int minTaps(int n, vector<int>& ranges) {
        vector<pii> intervals = {{inf,inf}};
        for (int i = 0; i <= n; i++)
            intervals.push_back({i-ranges[i],i+ranges[i]});
        sort(intervals.begin(), intervals.end());
        priority_queue<int,vector<int>> pq;
        int now = 0, cnt = 0;
        for (auto [s,e] : intervals){
            
            if (s > now){
                now = pq.top(); pq.pop();
                cnt++;
            }
            // cout << s << " " << e << " " << now << endl;
            if (now >= n)
                return cnt;
            if (s > now)
                return -1;
            pq.push(e);
        }
        return -1;
    }
};



using pii = pair<int,int>;
const int inf = 0x3f3f3f3f;

class Solution {
public:
    int minTaps(int n, vector<int>& ranges) {
        vector<pii> intervals = {{inf,inf}};
        for (int i = 0; i <= n; i++)
            intervals.push_back({i-ranges[i],i+ranges[i]});
        sort(intervals.begin(), intervals.end());
        int reach = 0, now = 0, cnt = 0;
        for (auto [s,e] : intervals){
            
            if (s > now){
                now = reach;
                cnt++;
            }
            // cout << s << " " << e << " " << now << endl;
            if (now >= n)
                return cnt;
            if (s > now)
                return -1;
            reach = max(reach, e);
        }
        return -1;
    }
};